### ISSUE TITLE
Issues with writing views in quick succession to the same github file

### ISSUE THREAD

**Issue by K-Meech** [mobie/mobie-viewer-fiji#278]
There seems to be a delay between changing files on github, and these changes being available through the api.

This means that adding a view to a file on github (through the UI), and then rapidly trying to add another, leads to issues. The old version is read, and hence you only end up with one of the views being written.

Not sure if this is something we can fix on our end... If not, we should warn about it in the docs.

**Comment by constantinpape** [comment]
Yes, github caches the raw view and only updates it every 5 minutes: https://stackoverflow.com/questions/62785962/get-raw-file-from-github-without-waiting-for-5-minute-cache-update.

> This means that adding a view to a file on github (through the UI), and then rapidly trying to add another, leads to issues. The old version is read, and hence you only end up with one of the views being written.
> 
> Not sure if this is something we can fix on our end... If not, we should warn about it in the docs.

How does adding a bookmark on github currently work? Does it directly write to main/master? That's maybe not such a good idea, because it can also lead to conflicts if multiple people write at the same time.
Maybe we could enforce that this happens on a separate branch and then one would need to make a PR to merge into main/master?

**Comment by K-Meech** [comment]
Thanks for the info! Makes sense - we should add a quick note to mobie.github.io on the project creation bit about this.

Yes - at the moment it commits directly to master (providing you have an access token with push access). I agree, it could lead to some conflicts, especially with the 5 min delay. 
I guess each added bookmark could create a branch from master, and add to that branch? Then if you add multiple bookmarks you end up with one branch for each? (could end up getting messy though if you add a lot of bookmarks...)
Or we have a certain branch name reserved for adding bookmarks, and they all get added to that one branch?

**Comment by constantinpape** [comment]
> Or we have a certain branch name reserved for adding bookmarks, and they all get added to that one branch?

I think it would be best to have one branch per user (we should be able to identify them by the token I hope.)

**Comment by K-Meech** [comment]
True - that should minimise conflicts as much as possible. At least then, any issues with the 5 min delay and writing multiple bookmarks is confined to an individual user's branch.

**Comment by constantinpape** [comment]
> At least then, any issues with the 5 min delay and writing multiple bookmarks is confined to an individual user's branch.

To solve this you would somehow need to introduce a local version of the bookmarks that buffer the version saved on github.
E.g. initialise the local version with the github bookmarks when creating the project, do all changes locally and sync with the github version every couple of minutes (>5).

**Comment by K-Meech** [comment]
Yes - that would work! It would need quite a bit of refactoring though.
Hmm... I'll look into this after all the project creation stuff is merged

**Comment by constantinpape** [comment]
Yes, I agree this is not highest priority.

**Comment by tischi** [comment]
> Yes, github caches the raw view and only updates it every 5 minutes

We should also change our code at some point to not rely on the "raw files" but fetch files using the github API.
I assume this would get rid of the delay. I made for this https://github.com/mobie/mobie-viewer-fiji/issues/279
